lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/CatalogFunctionAccessControlRule.java

```diff
@@ -0,0 +1,105 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.base.security;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.plugin.base.security.FunctionAccessControlRule.FunctionPrivilege;
+import io.trino.spi.connector.CatalogSchemaRoutineName;
+import io.trino.spi.function.FunctionKind;
+
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static java.util.Objects.requireNonNull;
+
+public class CatalogFunctionAccessControlRule
+{
+    public static final CatalogFunctionAccessControlRule ALLOW_ALL = new CatalogFunctionAccessControlRule(
+            FunctionAccessControlRule.ALLOW_ALL,
+            Optional.empty());
+
+    private final FunctionAccessControlRule functionAccessControlRule;
+    private final Optional<Pattern> catalogRegex;
+
+    @JsonCreator
+    public CatalogFunctionAccessControlRule(
+            @JsonProperty("privileges") Set<FunctionPrivilege> privileges,
+            @JsonProperty("user") Optional<Pattern> userRegex,
+            @JsonProperty("role") Optional<Pattern> roleRegex,
+            @JsonProperty("group") Optional<Pattern> groupRegex,
+            @JsonProperty("schema") Optional<Pattern> schemaRegex,
+            @JsonProperty("function_kinds") @JsonAlias("functionKinds") Set<FunctionKind> functionKinds,
+            @JsonProperty("function") Optional<Pattern> tableFunctionRegex,
+            @JsonProperty("catalog") Optional<Pattern> catalogRegex)
+    {
+        this(new FunctionAccessControlRule(privileges, userRegex, roleRegex, groupRegex, schemaRegex, functionKinds, tableFunctionRegex), catalogRegex);
+    }
+
+    private CatalogFunctionAccessControlRule(FunctionAccessControlRule functionAccessControlRule, Optional<Pattern> catalogRegex)
+    {
+        this.functionAccessControlRule = requireNonNull(functionAccessControlRule, "tableFunctionAccessControlRule is null");
+        this.catalogRegex = requireNonNull(catalogRegex, "catalogRegex is null");
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, String functionName)
+    {
+        return functionAccessControlRule.matches(user, roles, groups, functionName);
+    }
+
+    public boolean matches(String user, Set<String> roles, Set<String> groups, FunctionKind functionKind, CatalogSchemaRoutineName functionName)
+    {
+        if (!catalogRegex.map(regex -> regex.matcher(functionName.getCatalogName()).matches()).orElse(true)) {
+            return false;
+        }
+        return functionAccessControlRule.matches(user, roles, groups, functionKind, functionName.getSchemaRoutineName());
+    }
+
+    Optional<AnyCatalogPermissionsRule> toAnyCatalogPermissionsRule()
+    {
+        if (functionAccessControlRule.getPrivileges().isEmpty() || !functionAccessControlRule.isTableFunction()) {
```

Comment:
I introduce this to avoid reveal all catalogs and schemas - scalar, aggregate and window functions are not (at least for now) tied to any connector so adding any rule related to those function kinds cause revealing all catalogs and schemas.
cc @kokosing

Reply:
as discussed this lgtm.
I'd at least change of the order of sides of || condition.
I know you are not keen to invert this whole condition :)

Reply:
Please add a comment. (Same for other places like that)

Reply:
Added